File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.cpp
class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        int res = 0;
        int numUsed = 0;
        unordered_map<int, int> freqs;
        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> finishTimes;
        priority_queue<int, vector<int>, greater<int>> availableRooms;
        sort(meetings.begin(), meetings.end());
        for(int i = 0; i < n; i++)
        {
            availableRooms.push(i);
        }
        for(int i = 0; i < meetings.size(); i++)
        {
            while(!finishTimes.empty() && meetings[i][0] >= finishTimes.top().first)
            {
                availableRooms.push(finishTimes.top().second);
                finishTimes.pop();
            }
            int roomUsed;
            int usedRoomFreq;
            if(availableRooms.empty())
            {
                pair<long long, int> lastMeeting = finishTimes.top();
                finishTimes.pop();
                finishTimes.push({meetings[i][1] + lastMeeting.first - meetings[i][0], lastMeeting.second});
                freqs[lastMeeting.second]++;
                roomUsed = lastMeeting.second;
                usedRoomFreq = freqs[lastMeeting.second];
            }
            else
            {
                finishTimes.push({meetings[i][1], availableRooms.top()});
                freqs[availableRooms.top()]++;
                roomUsed = availableRooms.top();
                usedRoomFreq = freqs[availableRooms.top()];
                availableRooms.pop();
            }
            if(usedRoomFreq > numUsed)
            {
                res = roomUsed;
                numUsed = usedRoomFreq;
            }
            else if(usedRoomFreq == numUsed)
            {
                res = min(res, roomUsed);
            }
        }
        return res;
    }
};

File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.py
class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
        res = 0
        numUsed = 0
        freqs = {i:0 for i in range(n)}
        finishTimes = []
        availableRooms = [i for i in range(n)]
        meetings = sorted(meetings)
        for m in meetings:
            while len(finishTimes) > 0 and m[0] >= finishTimes[0][0]:
                heapq.heappush(availableRooms, finishTimes[0][1])
                heapq.heappop(finishTimes)
            if len(availableRooms) == 0:
                lastMeeting = finishTimes[0]
                heapq.heappop(finishTimes)
                heapq.heappush(finishTimes, (m[1] + lastMeeting[0] - m[0], lastMeeting[1]))
                freqs[lastMeeting[1]] += 1
                roomUsed = lastMeeting[1]
                usedRoomFreq = freqs[lastMeeting[1]]
            else:
                heapq.heappush(finishTimes, (m[1], availableRooms[0]))
                freqs[availableRooms[0]] += 1
                roomUsed = availableRooms[0]
                usedRoomFreq = freqs[availableRooms[0]]
                heapq.heappop(availableRooms)
            if usedRoomFreq > numUsed:
                res = roomUsed
                numUsed = usedRoomFreq
            elif usedRoomFreq == numUsed:
                res = min(res, roomUsed)
        return res

